pkg/server/handler/device.go

```diff
@@ -167,3 +182,90 @@ func (h *DeviceRegisterHandler) Handle(rpayload *router.Payload, response *route
 		response.Result = DeviceReigsterResult{device.ID}
 	}
 }
+
+// DeviceUnregisterHandler removes user id from a device
+//
+// Example to unregister a device:
+//
+//	curl -X POST -H "Content-Type: application/json" \
+//	  -d @- http://localhost:3000/ <<EOF
+//	{
+//		"action": "device:unregister",
+//		"access_token": "some-access-token",
+//		"type": "ios",
+//		"device_token": "some-device-token"
+//	}
+//	EOF
+//
+type DeviceUnregisterHandler struct {
+	Authenticator router.Processor `preprocessor:"authenticator"`
+	DBConn        router.Processor `preprocessor:"dbconn"`
+	InjectUser    router.Processor `preprocessor:"inject_user"`
+	InjectDB      router.Processor `preprocessor:"inject_db"`
+	RequireUser   router.Processor `preprocessor:"require_user"`
+	PluginReady   router.Processor `preprocessor:"plugin_ready"`
+	preprocessors []router.Processor
+}
+
+func (h *DeviceUnregisterHandler) Setup() {
+	h.preprocessors = []router.Processor{
+		h.Authenticator,
+		h.DBConn,
+		h.InjectUser,
+		h.InjectDB,
+		h.RequireUser,
+		h.PluginReady,
+	}
+}
+
+func (h *DeviceUnregisterHandler) GetPreprocessors() []router.Processor {
+	return h.preprocessors
+}
+
+func (h *DeviceUnregisterHandler) Handle(rpayload *router.Payload, response *router.Response) {
+	payload := deviceUnregisterPayload{}
+	if err := payload.Decode(rpayload.Data); err != nil {
+		response.Err = err
+		return
+	}
+
+	conn := rpayload.DBConn
+
+	device := skydb.Device{}
+	if err := conn.GetDevice(payload.ID, &device); err != nil {
+		if err == skydb.ErrDeviceNotFound {
+			response.Err = skyerr.NewError(skyerr.ResourceNotFound, "Device not found")
+			return
+		}
+
+		log.WithFields(logrus.Fields{
+			"deviceID": payload.ID,
+			"err":      err,
+		}).Errorln("Fail to get device")
+
+		response.Err = skyerr.NewResourceFetchFailureErr("device", payload.ID)
+		return
+	}
+
+	// delete all all devices with the same token
+	if err := conn.DeleteDevicesByToken(device.Token, skydb.ZeroTime); err != nil {
```

Comment:
This piece of code seems doesn’t belongs here. Is this to make sure there are no devices with the same token? But doing this will also delete the device that we will be saving.

Wondering why and wondering if there is a better place where this piece of code belongs.